File: src/main/java/com/example/application/views/gridproview/GridProView.java
package com.example.application.views.gridproview;

import java.util.Arrays;
import java.util.ArrayList;
import java.util.List;

import com.vaadin.flow.component.orderedlayout.VerticalLayout;
import com.vaadin.flow.component.gridpro.GridPro;
import com.vaadin.flow.component.gridpro.GridProVariant;
import com.vaadin.flow.component.textfield.IntegerField;
import com.vaadin.flow.router.PageTitle;
import com.vaadin.flow.router.Route;

@PageTitle("Grid Pro View")
@Route(value = "grid-pro-view")
public class GridProView extends VerticalLayout {
    /**
     * Configure GridPro as follows:
     *
     * 1. Only allow editing hours when the task status is "in progress"
     * 2. Disallow changing task status and hours when the task is archived
     * 3. Only allow changing archived if the status is "done" or "canceled"
     */
    public GridProView() {
        GridPro<Task> grid = new GridPro<>();

        grid.addEditColumn(Task::getName).text(Task::setName).setHeader("Name");

        List<String> statusOptions = Arrays.asList("todo", "in progress",
                "done", "canceled");
        grid.addEditColumn(Task::getStatus)
                .select(Task::setStatus, statusOptions).setHeader("Status");

        IntegerField hoursField = new IntegerField();
        hoursField.setWidthFull();

        grid.addEditColumn(Task::getHours).custom(hoursField, Task::setHours)
                .setHeader("Hours spent");

        grid.addEditColumn(Task::getArchived).checkbox(Task::setArchived)
                .setHeader("Archived");

        grid.setAllRowsVisible(true);
        grid.setItems(tasks);
        grid.addThemeVariants(GridProVariant.LUMO_HIGHLIGHT_READ_ONLY_CELLS);
        add(grid);
    }

    private List<Task> tasks = new ArrayList<Task>(
            Arrays.asList(new Task("Build a prototype", "done", 8, true),
                    new Task("Conduct the research", "canceled", 0, true),
                    new Task("Create the design", "in progress", 10, false),
                    new Task("Implement the UI", "todo", 0, false),
                    new Task("Add integration tests", "todo", 0, false),
                    new Task("Deploy to production", "todo", 0, false)));

    public static class Task {
        private String name;
        private String status;
        private int hours;
        private boolean archived;

        public Task(String name, String status, int hours, boolean archived) {
            this.name = name;
            this.status = status;
            this.hours = hours;
            this.archived = archived;
        }

        public String getName() {
            return name;
        }

        public void setName(String name) {
            this.name = name;
        }

        public String getStatus() {
            return status;
        }

        public void setStatus(String status) {
            this.status = status;
        }

        public int getHours() {
            return hours;
        }

        public void setHours(int hours) {
            this.hours = hours;
        }

        public boolean getArchived() {
            return archived;
        }

        public void setArchived(boolean archived) {
            this.archived = archived;
        }
    }
}

File: frontend/views/grid-pro-view/grid-pro-view.ts
import { html } from 'lit';
import { customElement, state } from 'lit/decorators.js';
import { View } from '../view';
import '@vaadin/grid-pro';
import '@vaadin/grid-pro/vaadin-grid-pro-edit-column.js';
import '@vaadin/vertical-layout';

@customElement('grid-pro-view')
export class GridProView extends View {
  @state()
  statuses = ['todo', 'in progress', 'done', 'canceled'];

  @state()
  tasks = [
    { name: 'Build a prototype', hours: 8, status: 'done', archived: true},
    { name: 'Conduct the research', hours: 0, status: 'canceled', archived: true},
    { name: 'Create the design', hours: 10, status: 'in progress', archived: false},
    { name: 'Implement the UI', hours: 0, status: 'todo', archived: false},
    { name: 'Add integration tests', hours: 0, status: 'todo', archived: false},
    { name: 'Deploy to production', hours: 0, status: 'todo', archived: false},
  ];

  /**
   * Configure GridPro as follows:
   *
   * 1. Only allow editing hours when the task status is "in progress"
   * 2. Disallow changing task status and hours when the task is archived
   * 3. Only allow changing archived if the status is "done" or "canceled"
   */
  render() {
    return html`
      <vaadin-vertical-layout theme="padding">
        <vaadin-grid-pro .items="${this.tasks}" all-rows-visible>
          <vaadin-grid-pro-edit-column path="name"></vaadin-grid-pro-edit-column>
           <vaadin-grid-pro-edit-column
            path="status"
            editor-type="select"
            .editorOptions="${this.statuses}"
          ></vaadin-grid-pro-edit-column>
          <vaadin-grid-pro-edit-column path="hours" header="Hours spent"></vaadin-grid-pro-edit-column>
          <vaadin-grid-pro-edit-column path="archived" editor-type="checkbox"></vaadin-grid-pro-edit-column>
        </vaadin-grid-pro>
      </vaadin-vertical-layout>
    `;
  }
}
